### PR TITLE
[SPARK-41064][CONNECT][PYTHON] Implement `DataFrame.crosstab` and `DataFrame.stat.crosstab`

### PR DESCRIPTION
**Pull request by zhengruifeng**
### What changes were proposed in this pull request?
Implement `DataFrame.crosstab` and `DataFrame.stat.crosstab`

### Why are the changes needed?
for api coverage

### Does this PR introduce _any_ user-facing change?
yes, new api

### How was this patch tested?
added ut

### CODE REVIEW

**Comment by zhengruifeng** [comment]
cc @HyukjinKwon @cloud-fan @amaliujia

File: connector/connect/src/main/protobuf/spark/connect/relations.proto [hunk 1]
```diff
@@ -272,5 +273,13 @@ message StatFunction {
   message Summary {
     repeated string statistics = 1;
   }
+
+  // StatFunctions.crossTabulate
+  message CrossTabulate {
```
Comment (amaliujia) [inline connector/connect/src/main/protobuf/spark/connect/relations.proto:278]: Can we add a bit more comments on such proto? Basically proto is also API that someone will read to understand the contract.

Beam's implementation offers a good example of comments for proto: https://github.com/apache/beam/blob/master/model/pipeline/src/main/proto/org/apache/beam/model/pipeline/v1/beam_runner_api.proto

Comment (amaliujia) [reply connector/connect/src/main/protobuf/spark/connect/relations.proto:278]: I am thinking for such function we can document
1. what this function does
2. its parameters (e.g. range, examples, etc.)
3. some caveats (invalid usage, etc.)

File: connector/connect/src/test/scala/org/apache/spark/sql/connect/planner/SparkConnectProtoSuite.scala [hunk 2]
```diff
@@ -267,6 +267,12 @@ class SparkConnectProtoSuite extends PlanTest with SparkConnectPlanTest {
       sparkTestRelation.summary("count", "mean", "stddev"))
   }
 
+  test("Test crosstab") {
+    comparePlans(
+      connectTestRelation.crosstab("id", "name"),
+      sparkTestRelation.stat.crosstab("id", "name"))
```
Comment (amaliujia) [inline connector/connect/src/test/scala/org/apache/spark/sql/connect/planner/SparkConnectProtoSuite.scala:273]: Can we also add the `stat` layer to the DSL so DSL can have the same invoke chain: relation.stat.xxx?

File: connector/connect/src/main/protobuf/spark/connect/relations.proto (hunk 1 above)
Comment (zhengruifeng) [reply connector/connect/src/main/protobuf/spark/connect/relations.proto:278]: nice, will update

File: connector/connect/src/main/protobuf/spark/connect/relations.proto [hunk 3]
```diff
@@ -273,6 +274,21 @@ message StatFunction {
   message Summary {
     repeated string statistics = 1;
   }
+
```
Comment (cloud-fan) [inline connector/connect/src/main/protobuf/spark/connect/relations.proto:277]: shall we make the proto message a bit more flattened? It looks to me that the only shared part between `Summary` and `CrossTabulate` is the input relation. Shall we simply have 2 messages?
```
message StatsSummary {
  Relation input = 1;
  ...
}
message StatsCrossTabulate {
  Relation input = 1;
  ...
}
```

Comment (amaliujia) [reply connector/connect/src/main/protobuf/spark/connect/relations.proto:277]: Or a different strategy is shared a field named arguments but the server side will do more to differentiate the inputs for Summary, CrossTab, etc. This seems to be bad.

Separating into different messages makes sense with existing example about  Summary and CrossTab

Comment (zhengruifeng) [reply connector/connect/src/main/protobuf/spark/connect/relations.proto:277]: ok, let me flatten summary first

Comment (HyukjinKwon) [reply connector/connect/src/main/protobuf/spark/connect/relations.proto:277]: +1 to make it flattened or simpler.

File: python/pyspark/sql/connect/dataframe.py [hunk 4]
```diff
@@ -454,3 +501,18 @@ def explain(self) -> str:
             return self._session.explain_string(query)
         else:
             return ""
+
+
+class DataFrameStatFunctions:
+    """Functionality for statistic functions with :class:`DataFrame`.
+
+    .. versionadded:: 3.4
```
Comment (HyukjinKwon) [inline python/pyspark/sql/connect/dataframe.py:509]:
```suggestion
    .. versionadded:: 3.4.0
```

File: connector/connect/src/main/scala/org/apache/spark/sql/connect/dsl/package.scala [hunk 5]
```diff
@@ -227,6 +227,21 @@ package object dsl {
       }
     }
 
+    implicit class DslStatFunctions(val logicalPlan: Relation) {
```
Comment (cloud-fan) [inline connector/connect/src/main/scala/org/apache/spark/sql/connect/dsl/package.scala:230]: does this need to be an implicit class?

**Review by cloud-fan — APPROVED** [review]

File: connector/connect/src/main/scala/org/apache/spark/sql/connect/dsl/package.scala (hunk 5 above)
Comment (zhengruifeng) [reply connector/connect/src/main/scala/org/apache/spark/sql/connect/dsl/package.scala:230]: seems not needed, but widely used in this file and [sql/catalyst/dsl/package](https://github.com/apache/spark/blob/ff7ab34a6957965d74504ed1d36de21d1ad7319c/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/dsl/package.scala#L374-L386)

**Comment by zhengruifeng** [comment]
merged into master, thank you all for reviews

**Comment by amaliujia** [comment]
Thanks. Late LGTM

**Pull request closed without merging**